lib/improver/tests/spotdata_new/spotdata_new/test_NeighbourSelection.py

```diff
@@ -0,0 +1,641 @@
+        result, result_nodes = plugin.build_KDTree(self.region_land_mask)
+        expected_length = np.nonzero(self.region_land_mask.data)[0].shape[0]
+
+        self.assertEqual(result_nodes.shape[0], expected_length)
+        self.assertIsInstance(result, scipy.spatial.ckdtree.cKDTree)
+
+
+class Test_select_minimum_dz(Test_NeighbourSelection):
+
+    """Test extraction of the minimum height difference points from a provided
+    array of neighbours."""
+
+    def test_basic(self):
+        """Test a simple case where the first element in the provided lists
+        has the smallest vertical displacement to the site. Expect the
+        coordinates of the first node to be returned."""
+
+        plugin = NeighbourSelection()
+        site_altitude = 3.
+        nodes = np.array([[0, 0], [1, 0], [2, 0], [3, 0], [4, 0]])
+        distance = np.arange(5)
+        indices = np.arange(5)
+
+        result = plugin.select_minimum_dz(self.region_orography,
+                                          site_altitude, nodes,
+                                          distance, indices)
+        self.assertArrayEqual(result, nodes[0])
+
+    def test_some_invalid_points(self):
+        """Test a case where some nodes are beyond the imposed search_radius,
+        which means they have a distance of np.inf, ensuring this is handled.
+        Also change the site height so the second node is the expected
+        result."""
+
+        plugin = NeighbourSelection()
+        site_altitude = 5.
+        nodes = np.array([[0, 0], [1, 0], [2, 0], [3, 0], [4, 0]])
+        distance = np.array([0, 1, 2, 3, np.inf])
+        indices = np.arange(5)
+
+        result = plugin.select_minimum_dz(self.region_orography,
+                                          site_altitude, nodes,
+                                          distance, indices)
+        self.assertArrayEqual(result, nodes[1])
+
+    def test_all_invalid_points(self):
+        """Test a case where all nodes are beyond the imposed search_radius,
+        so the returned value should be None."""
+
+        plugin = NeighbourSelection()
+        site_altitude = 5.
+        nodes = np.array([[0, 0], [1, 0], [2, 0], [3, 0], [4, 0]])
+        distance = np.full(5, np.inf)
+        indices = np.arange(5)
+
+        result = plugin.select_minimum_dz(self.region_orography,
+                                          site_altitude, nodes,
+                                          distance, indices)
+        self.assertEqual(result, None)
+
+
+class Test_process(Test_NeighbourSelection):
+
+    """Test the process method of the NeighbourSelection class."""
+
+    def test_global_attribute(self):
+        """Test that a cube is returned with an attribute identifying the
+        grid domain."""
+
+        plugin = NeighbourSelection()
+        result = plugin.process(self.global_sites, self.global_orography,
+                                self.global_land_mask)
+
+        expected = {'mosg__grid_domain': 'global'}
```

Comment:
This works fine as a test but grid_domain global does not uniquely define the grid for global data with mosg. gl_ens and gl_det have different grids.

Reply:
I have added the other mosg_ information that comes from StaGE just to be complete. These extra bits do uniquely define the grid.